Test failures with Python 3.12 and astropy 5.3.2

### Description

I'm building astropy 5.3.2 with python 3.12.0 rc1 inside the Fedora infrastructure and I'm getting errors in the following tests:

astropy/io/votable/tests/vo_test.py 
astropy/nddata/mixins/tests/test_ndslicing.py 
astropy/table/tests/test_table.py

```
=================================== FAILURES ===================================
____________________________ test_gzip_filehandles _____________________________

tmp_path = PosixPath('/tmp/pytest-of-mockbuild/pytest-0/test_gzip_filehandles0')

    def test_gzip_filehandles(tmp_path):
        with np.errstate(over="ignore"):
            # https://github.com/astropy/astropy/issues/13341
            votable = parse(get_pkg_data_filename("data/regression.xml"))
    
        # W39: Bit values can not be masked
        with pytest.warns(W39):
            with open(tmp_path / "regression.compressed.xml", "wb") as fd:
                votable.to_xml(fd, compressed=True, _astropy_version="testing")
    
        with open(tmp_path / "regression.compressed.xml", "rb") as fd:
>           votable = parse(fd)

astropy/io/votable/tests/vo_test.py:855: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
astropy/utils/decorators.py:604: in wrapper
    return function(*args, **kwargs)
astropy/io/votable/table.py:177: in parse
    return tree.VOTableFile(config=config, pos=(1, 1)).parse(iterator, config)
astropy/io/votable/tree.py:3858: in parse
    for start, tag, data, pos in iterator:
/usr/lib64/python3.12/gzip.py:324: in read
    return self._buffer.read(size)
/usr/lib64/python3.12/_compression.py:68: in readinto
    data = self.read(len(byte_view))
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <gzip._GzipReader object at 0x7f75bd008730>, size = 8192

    def read(self, size=-1):
        if size < 0:
            return self.readall()
        # size=0 is special because decompress(max_length=0) is not supported
        if not size:
            return b""
    
        # For certain input data, a single
        # call to decompress() may not return
        # any data. In this case, retry until we get some data or reach EOF.
        while True:
            if self._decompressor.eof:
                # Ending case: we've come to the end of a member in the file,
                # so finish up this member, and read a new gzip header.
                # Check the CRC and file size, and set the flag so we read
                # a new member
                self._read_eof()
                self._new_member = True
                self._decompressor = self._decomp_factory(
                    **self._decomp_args)
    
            if self._new_member:
                # If the _new_member flag is set, we have to
                # jump to the next member, if there is one.
                self._init_read()
                if not self._read_gzip_header():
                    self._size = self._pos
                    return b""
                self._new_member = False
    
            # Read a chunk of data from the file
            if self._decompressor.needs_input:
                buf = self._fp.read(READ_BUFFER_SIZE)
                uncompress = self._decompressor.decompress(buf, size)
            else:
                uncompress = self._decompressor.decompress(b"", size)
    
            if self._decompressor.unused_data != b"":
                # Prepend the already read bytes to the fileobj so they can
                # be seen by _read_eof() and _read_gzip_header()
                self._fp.prepend(self._decompressor.unused_data)
    
            if uncompress != b"":
                break
            if buf == b"":
>               raise EOFError("Compressed file ended before the "
                               "end-of-stream marker was reached")
E               EOFError: Compressed file ended before the end-of-stream marker was reached

/usr/lib64/python3.12/gzip.py:544: EOFError
_______________________ test_slicing_all_something_wrong _______________________

    def test_slicing_all_something_wrong():
        data = np.arange(10)
        mask = [False] * 10
        uncertainty = UnknownUncertainty({"rdnoise": 2.9, "gain": 1.4})
        naxis = 1
        wcs = nd_testing._create_wcs_simple(
            naxis=naxis,
            ctype=["deg"] * naxis,
            crpix=[3] * naxis,
            crval=[10] * naxis,
            cdelt=[1] * naxis,
        )
    
        nd = NDDataSliceable(data, mask=mask, uncertainty=uncertainty, wcs=wcs)
>       nd2 = nd[2:5]

astropy/nddata/mixins/tests/test_ndslicing.py:171: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
astropy/nddata/mixins/ndslicing.py:64: in __getitem__
    kwargs = self._slice(item)
astropy/nddata/mixins/ndslicing.py:92: in _slice
    kwargs["uncertainty"] = self._slice_uncertainty(item)
astropy/nddata/mixins/ndslicing.py:104: in _slice_uncertainty
    return self.uncertainty[item]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = UnknownUncertainty({'rdnoise': 2.9, 'gain': 1.4})
item = slice(2, 5, None)

    def __getitem__(self, item):
        """Normal slicing on the array, keep the unit and return a reference."""
>       return self.__class__(self.array[item], unit=self.unit, copy=False)
E       KeyError: slice(2, 5, None)

astropy/nddata/nduncertainty.py:331: KeyError
__________________________ test_table_attribute_fail ___________________________

    def test_table_attribute_fail():
        # Code raises ValueError(f'{attr} not allowed as TableAttribute') but in this
        # context it gets re-raised as a RuntimeError during class definition.
        with pytest.raises(RuntimeError, match="Error calling __set_name__"):
    
>           class MyTable2(Table):

astropy/table/tests/test_table.py:2900: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <[AttributeError("'TableAttribute' object has no attribute 'name'") raised in repr()] TableAttribute object at 0x7f75bb5f2cc0>
owner = <class 'astropy.table.tests.test_table.test_table_attribute_fail.<locals>.MyTable2'>
name = 'descriptions'

    def __set_name__(self, owner, name):
        import inspect
    
        params = [
            param.name
            for param in inspect.signature(owner).parameters.values()
            if param.kind
            not in (inspect.Parameter.VAR_KEYWORD, inspect.Parameter.VAR_POSITIONAL)
        ]
    
        # Reject names from existing params or best guess at parent class
        if name in params or hasattr(owner.__mro__[1], name):
>           raise ValueError(f"{name} not allowed as {self.__class__.__name__}")
E           ValueError: descriptions not allowed as TableAttribute
E           Error calling __set_name__ on 'TableAttribute' instance 'descriptions' in 'MyTable2'

astropy/utils/metadata.py:537: ValueError
```

### Expected behavior

Tests should pass

### How to Reproduce

Build with python 3.12


### Versions


Python 3.12.0rc1 (main, Aug  7 2023, 00:00:00) [GCC 13.2.1 20230728 (Red Hat 13.2.1-1)]

encodings: sys: utf-8, locale: utf-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.24.4
Scipy: 1.11.1
Matplotlib: 3.7.2
h5py: 3.8.0
Pandas: 1.5.3
PyERFA: 2.0.0.3
Cython: 3.0.0
Scikit-image: 0.21.0

--- a/.github/workflows/ci_workflows.yml
+++ b/.github/workflows/ci_workflows.yml
@@ -106,8 +106,9 @@ jobs:
           - language-pack-de
           - tzdata
       envs: |
-        - name: (Allowed Failure) Python 3.11 with remote data and dev version of key dependencies
-          linux: py311-test-devdeps
+        - name: (Allowed Failure) Python 3.12 with remote data and dev version of key dependencies
+          linux: py312-test-devdeps
+          python-version: '3.12-dev'
           posargs: --remote-data=any --verbose
 
         # https://github.com/matplotlib/matplotlib/issues/26847

--- a/.github/workflows/publish.yml
+++ b/.github/workflows/publish.yml
@@ -55,16 +55,19 @@ jobs:
         - cp39-manylinux_x86_64
         - cp310-manylinux_x86_64
         - cp311-manylinux_x86_64
+        - cp312-manylinux_x86_64
 
         # Note that following wheels are not currently tested:
 
         - cp39-manylinux_aarch64
         - cp310-manylinux_aarch64
         - cp311-manylinux_aarch64
+        - cp312-manylinux_aarch64
 
         - cp39-musllinux_x86_64
         - cp310-musllinux_x86_64
         - cp311-musllinux_x86_64
+        - cp312-musllinux_x86_64
 
         # MacOS X wheels - as noted in https://github.com/astropy/astropy/pull/12379 we deliberately
         # do not build universal2 wheels. Note that the arm64 wheels are not actually tested so we
@@ -73,20 +76,24 @@ jobs:
         - cp39*macosx_x86_64
         - cp310*macosx_x86_64
         - cp311*macosx_x86_64
+        - cp312*macosx_x86_64
 
         - cp39*macosx_arm64
         - cp310*macosx_arm64
         - cp311*macosx_arm64
+        - cp312*macosx_arm64
 
         # Windows wheels
 
         - cp39*win32
         - cp310*win32
         - cp311*win32
+        - cp312*win32
 
         - cp39*win_amd64
         - cp310*win_amd64
         - cp311*win_amd64
+        - cp312*win_amd64
 
     secrets:
       pypi_token: ${{ secrets.pypi_token }}

--- a/astropy/nddata/mixins/ndslicing.py
+++ b/astropy/nddata/mixins/ndslicing.py
@@ -105,8 +105,9 @@ class NDSlicingMixin:
             return None
         try:
             return self.uncertainty[item]
-        except TypeError:
+        except (TypeError, KeyError):
             # Catching TypeError in case the object has no __getitem__ method.
+            # Catching KeyError for Python 3.12.
             # But let IndexError raise.
             log.info("uncertainty cannot be sliced.")
         return self.uncertainty
@@ -116,7 +117,7 @@ class NDSlicingMixin:
             return None
         try:
             return self.mask[item]
-        except TypeError:
+        except (TypeError, KeyError):
             log.info("mask cannot be sliced.")
         return self.mask
 

--- a/astropy/table/tests/test_table.py
+++ b/astropy/table/tests/test_table.py
@@ -2898,14 +2898,19 @@ def test_table_attribute_ecsv():
 
 
 def test_table_attribute_fail():
-    # Code raises ValueError(f'{attr} not allowed as TableAttribute') but in this
-    # context it gets re-raised as a RuntimeError during class definition.
-    with pytest.raises(RuntimeError, match="Error calling __set_name__"):
+    if sys.version_info[:2] >= (3, 12):
+        ctx = pytest.raises(ValueError, match=".* not allowed as TableAttribute")
+    else:
+        # Code raises ValueError(f'{attr} not allowed as TableAttribute') but in this
+        # context it gets re-raised as a RuntimeError during class definition.
+        ctx = pytest.raises(RuntimeError, match="Error calling __set_name__")
+
+    with ctx:
 
         class MyTable2(Table):
             descriptions = TableAttribute()  # Conflicts with init arg
 
-    with pytest.raises(RuntimeError, match="Error calling __set_name__"):
+    with ctx:
 
         class MyTable3(Table):
             colnames = TableAttribute()  # Conflicts with built-in property

--- a/docs/changes/14784.other.rst
+++ b/docs/changes/14784.other.rst
@@ -1,0 +1,1 @@
+Compatibility with Python 3.12.

--- a/docs/time/index.rst
+++ b/docs/time/index.rst
@@ -512,7 +512,7 @@ Example
 
 To infer input format::
 
-  >>> from datetime import datetime
+  >>> from datetime import datetime, timezone
   >>> t = Time(datetime(2010, 1, 2, 1, 2, 3))
   >>> t.format
   'datetime'
@@ -717,7 +717,7 @@ The current time can be determined as a |Time| object using the
 `~astropy.time.Time.now` class method::
 
   >>> nt = Time.now()
-  >>> ut = Time(datetime.utcnow(), scale='utc')
+  >>> ut = Time(datetime.now(tz=timezone.utc), scale='utc')
 
 The two should be very close to each other.
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -202,6 +202,10 @@ doctest_norecursedirs = [
     "*/tests/command.py",
 ]
 doctest_subpackage_requires = [
+    "astropy/cosmology/_io/mapping.py = python<3.12",  # PYTHON_LT_3_12 (PR 14784)
+    "astropy/cosmology/_io/row.py = python<3.12",  # PYTHON_LT_3_12 (PR 14784)
+    "astropy/cosmology/_io/table.py = python<3.12",  # PYTHON_LT_3_12 (PR 14784)
+    "astropy/table/table.py = python<3.12",  # PYTHON_LT_3_12 (PR 14784)
     "astropy/table/mixins/dask.py = dask",
     "docs/io/fits/index.rst = numpy<1.25",  # NUMPY_LT_1_25 (Issue 14545)
     "docs/io/fits/usage/image.rst = numpy<1.25",  # NUMPY_LT_1_25 (Issue 14545)

--- a/tox.ini
+++ b/tox.ini
@@ -1,6 +1,6 @@
 [tox]
 envlist =
-    py{39,310,311,dev}-test{,-recdeps,-alldeps,-oldestdeps,-devdeps,-mpldev,-devinfra,-predeps,-numpy122,-numpy123,-mpl334}{,-cov}{,-clocale}{,-fitsio}
+    py{39,310,311,312,dev}-test{,-recdeps,-alldeps,-oldestdeps,-devdeps,-mpldev,-devinfra,-predeps,-numpy122,-numpy123,-mpl334}{,-cov}{,-clocale}{,-fitsio}
     # Only these two exact tox environments have corresponding figure hash files.
     py39-test-image-mpl334-cov
     py39-test-image-mpldev-cov

